### PR TITLE
Blog app - integration specs for views and fix n+1 problems

### PR DESCRIPTION
**Pull request by Mikey1-nuel**
## Blog app - integration specs for views and fix n+1 problems
**In this pull request, the following tasks were implemented:**

- Made sure that the N+1 problem was solved when fetching all posts and their comments for a user.

Screenshot (463)
Screenshot (464)
Screenshot (465)
Screenshot (466)

- Used Capybara to write integration tests for each view in my project.

### CODE REVIEW

File: spec/system/user_show_spec.rb [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+    it "should show the user's bio" do
+      expect(page).to have_content user.bio
+    end
+
+    it "should show the user's first 3 posts" do
+      user.return_three_most_recent_posts.each do |post|
+        expect(page).to have_content post.title
+      end
+    end
```
Comment (danmainah) [inline spec/system/user_show_spec.rb]: - [ ] @Mikey1-nuel and @saluumaa  Great work on your tests for the user show page, however in this test, we were supposed to test if using the recent_posts method we can view only view user's first 3 posts, kindly consider modifying this test

File: spec/system/user_show_spec.rb [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+    it 'should display a button that lets me view all of a user\'s posts' do
+      expect(page).to have_link('View all posts')
+    end
+
+    it 'should redirect to the user\'s post\'s index page when I click to see all posts' do
+      click_link('View all posts')
+      expect(page).to have_current_path(user_posts_path(user))
+    end
+  end
```
Comment (danmainah) [inline spec/system/user_show_spec.rb:58]: - [ ] @team great work adding tests for the user show page, however, you forgot to add a test that asserts the user is redirected  to that post's show page when the user clicks on the post

File: spec/system/post_index_spec.rb [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+      end
+    end
+  end
+
+  describe 'clicking on post' do
+    it 'redirects to posts show page' do
+      click_link(href: user_post_path(user, first_post))
+      expect(page).to have_current_path(user_post_path(user, first_post))
+    end
+  end
```
Comment (danmainah) [inline spec/system/post_index_spec.rb:72]: - [ ] @team Great work adding tests for the post page however you forgot to add the test `I can see a section for pagination if there are more posts than fit on the view.` kindly consider adding it.

**Review by danmainah — CHANGES_REQUESTED** [review]
Hi @Mikey1-nuel and @saluumaa , 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights

✔️ Fixed the N + 1 problem
✔️ Added tests for all 4 pages

# Required Changes ♻️

- [ ] @Mikey1-nuel and @saluumaa Great work fixing N + 1 problem, however, you didn't provide the correct screenshot of the before and after the query, check the example below, you run the app and check the terminal outcome. 
image

- [ ] @Team kindly ensure all tests are passing after implementing the requested changes

_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

File: spec/system/user_show_spec.rb [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    end
+
+    it "should show the user's bio" do
+      expect(page).to have_content user.bio
+    end
+
+    it "should show the user's first 3 posts" do
+      user.return_three_most_recent_posts.each do |post|
+        expect(page).to have_content post.title
+      end
```
Comment (Ismailco) [inline spec/system/user_show_spec.rb]: - [ ] Kindly consider following the previous reviewer's suggestion, this test should assert that the `return_three_most_recent_posts` method returns only the last three posts, you can do that by checking the count of the returned values/posts (it should be only 3 posts) and the titles of the last three post should equal post.title, you can choose other ideas to test this method

File: spec/system/user_show_spec.rb (hunk 2 above)
Comment (Ismailco) [inline spec/system/user_show_spec.rb]: - [ ] Same here as the previous reviewer mentioned, kindly implement tests that redirect to the post's show page when you click on a user's post.

File: spec/system/post_index_spec.rb (hunk 3 above)
Comment (Ismailco) [inline spec/system/post_index_spec.rb:72]: - [ ] Same here as the previous reviewer mentioned, kindly implement tests that assert you can see a section for pagination if there are more posts than fit on the view.

**Review by Ismailco — CHANGES_REQUESTED** [review]
### Hi @Mikey1-nuel,

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Highlights

✔️ No linter errors
✔️ Used Gitflow
✔️ Work documented professionally

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by cynthiainga — APPROVED** [review]
Hi @Mikey1-nuel, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 



## Highlights

- [ ] All tests are implemented and are passing ✔️ 
- [ ] No linters errors ✔️ 
- [ ] Descriptive PR ✔️ 

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were missed by the previous reviewer and addressing them will really improve your application._

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**